Refactor some enums in TypingCommand to enum classes

jesxilin opened the pull request:
#### 00fd2547e1128196f8adce915fa5f436caf86242
<pre>
Refactor some enums in TypingCommand to enum classes
<a href="https://bugs.webkit.org/show_bug.cgi?id=256936">https://bugs.webkit.org/show_bug.cgi?id=256936</a>
rdar://109488827

Reviewed by Wenson Hsieh.

Refactor enums ETypingCommand, TextCompositionType and Option in TypingCommand to be enum classes.
Changed instances of TypingCommand::Option options to OptionSet.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::deleteWithDirection):
(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::insertLineBreak):
(WebCore::Editor::insertParagraphSeparator):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeDelete):
(WebCore::executeInsertLineBreak):
(WebCore::executeInsertParagraph):
(WebCore::executeInsertText):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::editActionForTypingCommand):
(WebCore::TypingCommand::TypingCommand):
(WebCore::TypingCommand::deleteSelection):
(WebCore::TypingCommand::deleteKeyPressed):
(WebCore::TypingCommand::forwardDeleteKeyPressed):
(WebCore::TypingCommand::insertText):
(WebCore::TypingCommand::insertLineBreak):
(WebCore::TypingCommand::insertParagraphSeparatorInQuotedContent):
(WebCore::TypingCommand::insertParagraphSeparator):
(WebCore::TypingCommand::doApply):
(WebCore::TypingCommand::markMisspellingsAfterTyping):
(WebCore::TypingCommand::insertTextRunWithoutNewlines):
(WebCore::TypingCommand::updatePreservesTypingStyle):
* Source/WebCore/editing/TypingCommand.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::insertUnicodeCharacter):

Canonical link: <a href="https://commits.webkit.org/264493@main">https://commits.webkit.org/264493@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24ffa825a1e988e4eabc2c407d0325763b99e7e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8226 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9414 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7924 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7778 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10035 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7965 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10789 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7128 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9525 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6325 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14737 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7474 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10605 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7696 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6272 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7022 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1864 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11233 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7428 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -522,7 +522,7 @@ static bool executeInsertOrderedList(LocalFrame& frame, Event*, EditorCommandSou
 static bool executeInsertParagraph(LocalFrame& frame, Event*, EditorCommandSource, const String&)
 {
-    TypingCommand::insertParagraphSeparator(*frame.document(), 0);
+    TypingCommand::insertParagraphSeparator(*frame.document(), OptionSet<TypingCommand::Option> { });
```
this should be just `{ }` instead of `OptionSet<TypingCommand::Option> { }` (if it compiles)

rr-codes commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -533,7 +533,7 @@ static bool executeInsertTab(LocalFrame& frame, Event* event, EditorCommandSourc
 static bool executeInsertText(LocalFrame& frame, Event*, EditorCommandSource, const String& value)
 {
-    TypingCommand::insertText(*frame.document(), value, 0);
+    TypingCommand::insertText(*frame.document(), value, OptionSet<TypingCommand::Option> { });
```
ditto

rr-codes commented on Source/WebCore/editing/TypingCommand.cpp:
```diff
@@ -83,44 +83,44 @@ class TypingCommandLineOperation
-    if (compositionType == TypingCommand::TextCompositionPending) {
-        if (command == TypingCommand::InsertText)
+    if (compositionType == TypingCommand::TextCompositionType::Pending) {
+        if (command == TypingCommand::ETypingCommand::InsertText)
```
`ETypingCommand` should probably be renamed to just `Type` to reduce redundancy

rr-codes commented on Source/WebCore/editing/TypingCommand.h:
```diff
@@ -31,7 +31,7 @@ namespace WebCore {
 class TypingCommand final : public TextInsertionBaseCommand {
 public:
-    enum ETypingCommand { 
+    enum class ETypingCommand { // should be type
```
extra comment

pxlcoder replied:
`enum class ETypingCommand : uint8_t`

rr-codes commented on Source/WebCore/editing/TypingCommand.h:
```diff
@@ -41,29 +41,28 @@ class TypingCommand final : public TextInsertionBaseCommand {
     };
 
-    enum Option {
+    enum class Option : unsigned {
```
this should be `uint8_t` instead of `unsigned`

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -591,7 +591,7 @@ bool Editor::deleteWithDirection(SelectionDirection direction, TextGranularity g
     if (m_document.selection().isRange()) {
         if (isTypingAction) {
-            TypingCommand::deleteKeyPressed(document(), canSmartCopyOrDelete() ? TypingCommand::SmartDelete : 0, granularity);
+            TypingCommand::deleteKeyPressed(document(), canSmartCopyOrDelete() ? TypingCommand::Option::SmartDelete : OptionSet<TypingCommand::Option> { }, granularity);
```
maybe try `{ }` instead of `OptionSet<TypingCommand::Option> { }`? (or, if that doesn't work, `OptionSet { }` maybe)

jesxilin replied:
It seems that neither work, so I think I have to keep what was originally there.

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -1380,14 +1380,14 @@ bool Editor::insertTextWithoutSendingTextEvent(const String& text, bool selectIn
-                TypingCommand::insertText(document, text, selection, options, triggeringEvent && triggeringEvent->isComposition() ? TypingCommand::TextCompositionFinal : TypingCommand::TextCompositionNone);
+                    options.add(TypingCommand::Option::IsAutocompletion);
+                if (shouldRemoveAutocorrectionIndicator(shouldConsiderApplyingAutocorrection, autocorrectionWasApplied, options.contains(TypingCommand::Option::IsAutocompletion)))
+                    options.remove({ TypingCommand::Option::RetainAutocorrectionIndicator });
```
does this work if you remove the `{` and `}`?

jesxilin replied:
It does work!

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2154,9 +2154,9 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
     // Always delete the current composition before inserting the finalized composition text if we're confirming our composition.
     // Our default behavior (if the beforeinput event is not prevented) is to insert the finalized composition text back in.
-    // We pass TypingCommand::TextCompositionPending here to indicate that we are deleting the pending composition.
+    // We pass TypingCommand::Pending here to indicate that we are deleting the pending composition.
```
this should probably be `TypingCommand::TextCompositionType::Pending`

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2154,9 +2154,9 @@ void Editor::setComposition(const String& text, SetCompositionMode mode)
+    // We pass TypingCommand::Pending here to indicate that we are deleting the pending composition.
     if (mode != CancelComposition)
-        TypingCommand::deleteSelection(document(), 0, TypingCommand::TextCompositionPending);
+        TypingCommand::deleteSelection(document(), OptionSet<TypingCommand::Option> { }, TypingCommand::TextCompositionType::Pending);
```
I think the `OptionSet<TypingCommand::Option>` can be omitted

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2186,8 +2186,8 @@ void Editor::setComposition(const String& text, const Vector<CompositionUnderlin
     if (isStartingToRecomposeExistingRange) {
-        // We pass TypingCommand::TextCompositionFinal here to indicate that we are removing composition text that has been finalized.
-        TypingCommand::deleteSelection(document(), 0, TypingCommand::TextCompositionFinal);
+        // We pass TypingCommand::Final here to indicate that we are removing composition text that has been finalized.
```
ditto

rr-codes commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -2186,8 +2186,8 @@ void Editor::setComposition(const String& text, const Vector<CompositionUnderlin
-        // We pass TypingCommand::TextCompositionFinal here to indicate that we are removing composition text that has been finalized.
-        TypingCommand::deleteSelection(document(), 0, TypingCommand::TextCompositionFinal);
+        // We pass TypingCommand::Final here to indicate that we are removing composition text that has been finalized.
+        TypingCommand::deleteSelection(document(), OptionSet<TypingCommand::Option> { }, TypingCommand::TextCompositionType::Final);
```
I think the `OptionSet<TypingCommand::Option>` can be omitted

rr-codes commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -494,7 +494,7 @@ static bool executeInsertLineBreak(LocalFrame& frame, Event* event, EditorComman
         // InsertLineBreak is not implemented in IE or Firefox, so this behavior is only needed for
         // backward compatibility with ourselves, and for consistency with other commands.
-        TypingCommand::insertLineBreak(*frame.document(), 0);
+        TypingCommand::insertLineBreak(*frame.document(), OptionSet<TypingCommand::Option> { });
```
ditto

pxlcoder commented on Source/WebCore/editing/TypingCommand.h:
```diff
@@ -41,29 +41,28 @@ class TypingCommand final : public TextInsertionBaseCommand {
-        TextCompositionNone,
-        TextCompositionPending,
-        TextCompositionFinal,
+    enum class TextCompositionType {
```
`enum class TextCompositionType : uint8_t`

whsieh commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -600,11 +600,11 @@ bool Editor::deleteWithDirection(SelectionDirection direction, TextGranularity g
         }
     } else {
-        TypingCommand::Options options = 0;
+        OptionSet<TypingCommand::Option> options = { };
```
Nit - `OptionSet`'s default initializer will set all flags to 0, so this should be slightly cleaner:

```suggestion
        OptionSet<TypingCommand::Option> options;
```

jesxilin committed:
Refactor some enums in TypingCommand to enum classes
https://bugs.webkit.org/show_bug.cgi?id=256936
rdar://109488827

Reviewed by Wenson Hsieh.

Refactor enums ETypingCommand, TextCompositionType and Option in TypingCommand to be enum classes.
Changed instances of TypingCommand::Option options to OptionSet.

* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::deleteWithDirection):
(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::insertLineBreak):
(WebCore::Editor::insertParagraphSeparator):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeDelete):
(WebCore::executeInsertLineBreak):
(WebCore::executeInsertParagraph):
(WebCore::executeInsertText):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::editActionForTypingCommand):
(WebCore::TypingCommand::TypingCommand):
(WebCore::TypingCommand::deleteSelection):
(WebCore::TypingCommand::deleteKeyPressed):
(WebCore::TypingCommand::forwardDeleteKeyPressed):
(WebCore::TypingCommand::insertText):
(WebCore::TypingCommand::insertLineBreak):
(WebCore::TypingCommand::insertParagraphSeparatorInQuotedContent):
(WebCore::TypingCommand::insertParagraphSeparator):
(WebCore::TypingCommand::doApply):
(WebCore::TypingCommand::markMisspellingsAfterTyping):
(WebCore::TypingCommand::insertTextRunWithoutNewlines):
(WebCore::TypingCommand::updatePreservesTypingStyle):
* Source/WebCore/editing/TypingCommand.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::insertUnicodeCharacter):

Canonical link: https://commits.webkit.org/264493@main
